packages/adapter-begin/index.js

```diff
@@ -24,27 +22,31 @@ function parse_arc(arcPath) {
 	}
 }
 
-export default async function adapter(builder) {
-	builder.log.minor('Parsing app.arc file');
-	const { static: static_mount_point } = parse_arc('app.arc');
-
-	const lambda_directory = resolve(join('src', 'http', 'get-index'));
-	const static_directory = resolve(static_mount_point);
-	const server_directory = resolve(join('src', 'shared'));
-
-	builder.log.minor('Writing client application...');
-	builder.copy_static_files(static_directory);
-	builder.copy_client_files(static_directory);
-
-	builder.log.minor('Building lambda...');
-	const local_lambda_dir = join(__dirname, 'files');
-	copy(local_lambda_dir, lambda_directory);
-
-	builder.log.minor('Writing server application...');
-	builder.copy_server_files(server_directory);
-
-	builder.log.minor('Prerendering static pages...');
-	await builder.prerender({
-		dest: static_directory
-	});
+export default function adapter() {
+	return {
+		async adapt(builder) {
```

Comment:
It's a bit hard for me to envision the new API without a more explicit example of the use case we're trying to enable. One thing I wonder is whether we'd want `builder` passed in here or above in `adapter()`. If we add other methods that want to use it that could mean that we only have to pass it in once instead of for each method

Reply:
It's a big if, and I'm not sure I like it, because it's more likely that we'll want to instead pass other things, which would then be passed into the `adapter/createAdapter` function above, which would reasonably limit the number of things we pass before it all gets a bit silly.

I think pass it down to only where it is required until we find a reason not to do that.

Reply:
> It's a bit hard for me to envision the new API without a more explicit example of the use case we're trying to enable.

Think e.g.

```js
return {
  format: 'cjs', // because we're targeting lambda — see #503
  async deploy() {
    // do some dockery stuff or faff about with cache settings on your CDN
  },
  async adapt(builder) {
    // what we have now
  }
};
```



> One thing I wonder is whether we'd want `builder` passed in here or above in `adapter()`.

Should definitely be passed to `adapt`. If it needs to return other methods that would also use `builder` (unlikely, I think) then `builder` would be passed to those methods. Aside from it being unnecessary to pass `builder` in earlier (and harder, if we end up invoking this function before `adapt` happens, e.g. before `build` so that we know what format to emit), it means that you can't create adapters inline if we did #439.

What I mean by that is that you can create ad-hoc Rollup plugins like this...

```js
plugins: [
  {
    name: 'my-ad-hoc-transformation-plugin',
    transform(code, id) {
      // do some transformation that needn't live in a plugin package
    }
  }
]
```

...and it would be neat if you could do the same thing with adapters.